### PR TITLE
Add `Pool::add`

### LINKED ISSUE #212: [feature] Add/Get New Pooled Connections
I have a use case where I want to periodically re-evaluate my login credentials against the underlying database.

Using something like `Pool::get()` doesn't work, because it doesn't necessarily mean that the connection is new - it could be from the pool instead. Luckily, this is pretty trivial today with `Pool::dedicated_connection()`:

```rust
let pool = Pool::builder().build(mgr).await.unwrap();
let mut interval = time::interval(time::Duration::from_secs(5));
loop {
  tokio::select! {
    _ = interval.tick() => match pool.dedicated_connection().await {
      Ok(_) => (),
      Err(err) => handle_connection_error(err),
  }
}
```

However, establishing a connection against my database is expensive, and I don't want to just throw away the connection after its established. Instead, it would be great if that new connection could be added to the pool.

I see two obvious ways of going about this:

1. A method to _add_ a connection to the pool:

```rust
/// Adds a connection to the pool.
pub fn add(&self, conn: M::Connection) -> ()
```

```rust
match pool.dedicated_connection().await {
  Ok(conn) => pool.add(conn),
  Err(err) => handle_connection_error(err),
}
```

2. A method to get a _new_ connection from the pool:

```rust
/// Retrieves a new connection from the pool.
pub async fn get_new(&self) -> Result<PooledConnection<'_, M>, RunError<M::Error>>
```

```rust
match pool.get_new().await {
  Ok(_pooled_connection) => (), // returned to pool on drop
  Err(err) => handle_connection_error(err),
}
```

I'm leaning towards the first option, since it's both easier to implement and more flexible. For example, you can create connections independent from the pool and add them to be managed later.

I would be happy to write the code for this feature if it's something you're ok with adding to the project.

### PR DESCRIPTION
**Pull request by tneely**
Fixes #212

This adds `Pool::add`, which allows for externally created connections to be added and managed by the pool. If the pool is at maximum capacity when this method is called, it will return the input connection as part of the Err response.

I considered allowing `Pool:add` to ignore `max_size` when adding to the pool, but felt it could lead to confusion if the pool is allowed to exceed its capacity in this specific case.

This change required making PoolInternals::approvals visible within the crate to get the approval needed to add a new connection. The alternative would have required defining a new pub(crate) method for this specific use case, which feels worse. I'm open to suggestions on how to more cleanly integrate this change into the package.

### CODE REVIEW

File: bb8/src/inner.rs [hunk 1]
```diff
@@ -161,6 +161,20 @@ where
         }
     }
 
+    /// Adds an external connection to the pool if there is capacity for it.
+    pub(crate) fn try_put(&self, conn: M::Connection) -> Result<(), M::Connection> {
```
Comment (djc) [inline bb8/src/inner.rs:165]: It might be better to move this logic into a method on `SharedPool`, which would also allow us to keep `approvals()` private. Then, I think we want a layer here which would run `manager.has_broken()` on the new `conn` before we add it. Might need another error enum to accomodate the different error paths.

File: bb8/src/inner.rs [hunk 2]
```diff
@@ -161,6 +161,20 @@ where
...
+        let mut locked = self.inner.internals.lock();
+        let mut approvals = locked.approvals(&self.inner.statics, 1);
+        match approvals.next() {
```
Comment (djc) [inline bb8/src/inner.rs:168]: Suggest spelling this as `let Some(approval) = approvals.next() else { return Err(conn); }`.

File: bb8/src/inner.rs (hunk 1 above)
Comment (tneely) [reply bb8/src/inner.rs:165]: That makes sense! Let me move things around

File: bb8/src/inner.rs (hunk 2 above)
Comment (tneely) [reply bb8/src/inner.rs:168]: Sounds good!

**Comment by tneely** [comment]
Sorry for the delay - I lost track of this PR at some point. I've moved a few things around based on your feedback and also added an `AddError` enum to differentiate between when the connection is broken and when there's no capacity.

File: bb8/src/api.rs [hunk 3]
```diff
@@ -547,3 +555,36 @@ impl<E> ErrorSink<E> for NopErrorSink {
...
+pub enum AddError<C> {
+    /// The connection was broken before it could be added.
+    Broken(C),
+    /// Unable to add the connection to the pool due to insufficient capacity.
+    NoCapacity(C),
+}
+
+impl<E> fmt::Display for AddError<E>
+where
+    E: error::Error + 'static,
```
Comment (djc) [inline bb8/src/api.rs:570]: Nit: please put these bounds into the `impl` declaration for conciseness (so `impl<E: error::Error + 'static>`).

File: bb8/src/api.rs [hunk 4]
```diff
@@ -547,3 +555,36 @@ impl<E> ErrorSink<E> for NopErrorSink {
         Box::new(*self)
     }
 }
+
+/// Error type returned by `Pool::add(conn)`
+#[derive(Debug, Clone, PartialEq, Eq)]
```
Comment (djc) [inline bb8/src/api.rs:560]: Nit: let's move this next to `RunError` (so above `ErrorSink`).

**Review by djc — COMMENTED** [review]
Looking pretty good, some small suggestions left!

File: bb8/tests/test.rs [hunk 5]
```diff
@@ -104,6 +104,37 @@ where
...
+    _c: PhantomData<C>,
+}
+
+impl<C> BrokenConnectionManager<C> {
+    fn new() -> Self {
+        BrokenConnectionManager { _c: PhantomData }
+    }
+}
+
+#[async_trait]
```
Comment (djc) [inline bb8/tests/test.rs:117]: Sorry, missed this previously: I want to review this move separately, so please isolate it in a separate commit (and squash all the other commits) together.

**Review by djc — APPROVED** [review]
This mostly looks good, needs one tweak to the commit history.

File: bb8/tests/test.rs (hunk 5 above)
Comment (djc) [reply bb8/tests/test.rs:117]: The idea here is that I want the move of the existing code here isolated in a separate commit from adding the new test -- sorry if that wasn't clear from my previous comment.

Comment (tneely) [reply bb8/tests/test.rs:117]: I think I understand? Let me try again.

Comment (djc) [reply bb8/tests/test.rs]: Not quite, but I've taken the liberty of fixing it up.

**Comment by djc** [comment]
Thanks for your patience through a number of iterations!

**Comment by tneely** [comment]
You're welcome, thanks for working through the PR with me!
